### PR TITLE
Performance measurements: Add Global Performance Measurements

### PR DESCRIPTION
**Pull request by tobonex**
Adds global performance measurements feature. It measures performance of a limited number of module instances. This feature is disabled by default and enabled by an IPC message.

### CODE REVIEW

File: src/debug/telemetry/telemetry.c
Comment (lgirdwood) [inline src/debug/telemetry/telemetry.c:93]: Could some of the getters and setters be static inline ?

File: src/audio/component.c
Comment (lgirdwood) [inline src/audio/component.c:561]: this looks like a private API ? so could be static ?

File: src/include/sof/debug/telemetry/telemetry.h
Comment (lgirdwood) [inline src/include/sof/debug/telemetry/telemetry.h:20]: ack - we should also prefix with ```SOF_TELEMETRY_```

File: src/debug/telemetry/telemetry.c
Comment (lgirdwood) [inline src/debug/telemetry/telemetry.c:170]: is this worthy to log ? i.e will it help with debug ?

File: src/debug/telemetry/telemetry.c
Comment (lgirdwood) [inline src/debug/telemetry/telemetry.c:178]: could do with a comment explaining this block

File: src/debug/telemetry/telemetry.c
Comment (lgirdwood) [inline src/debug/telemetry/telemetry.c:194]: ditto re comment

File: src/debug/telemetry/telemetry.c
Comment (lgirdwood) [inline src/debug/telemetry/telemetry.c:255]: could do with a comment for this block

File: src/debug/telemetry/telemetry.c
Comment (lgirdwood) [inline src/debug/telemetry/telemetry.c:306]: I guess can be removed ?

File: src/debug/telemetry/telemetry.c
Comment (lgirdwood) [inline src/debug/telemetry/telemetry.c:328]: comment explaining block

**Review by lgirdwood — COMMENTED** [review]
Quick look at a high level, nothing major. Could do with more comments around the larger code blocks.

File: src/debug/telemetry/telemetry.c
Comment (lyakh) [inline src/debug/telemetry/telemetry.c:101]: this function returns `int`, not `bool`. Same below

File: src/audio/component.c
Comment (lyakh) [inline src/audio/component.c:565]: external parentheses

File: src/audio/component.c
Comment (lyakh) [inline src/audio/component.c:568]: as @lgirdwood suggested - if it's `inline`, you probably meant it to be `static`. I see only 2 functions being added to a header, the rest presumably can be `static`. And I wouldn't use `inline` in .c

File: src/audio/component.c
Comment (lyakh) [inline src/audio/component.c:585]: commented out?

File: src/include/sof/audio/component.h
Comment (lyakh) [inline src/include/sof/audio/component.h:619]: why ending with an underscore? Don't think this is common practice in SOF

File: src/audio/component.c
Comment (lyakh) [inline src/audio/component.c:603]: ?

File: src/audio/component.c
Comment (lyakh) [inline src/audio/component.c:519]: where should this macro be defined? Should it be a Kconfig option?

File: src/audio/component.c
Comment (lyakh) [inline src/audio/component.c:532]: SOF doesn't use braces for single-line `if` and similar

File: src/debug/telemetry/telemetry.c
Comment (lyakh) [inline src/debug/telemetry/telemetry.c:41]: `volatile`? Maybe `static` - for as many of them as possible? And the terminating underscore

File: src/debug/telemetry/telemetry.c
Comment (lyakh) [inline src/debug/telemetry/telemetry.c:142]: if this is supposed to return POSIX errors, just `return ret;` and `return 0` below

File: src/debug/telemetry/telemetry.c
Comment (lyakh) [inline src/debug/telemetry/telemetry.c:281]: same here, mixed up return codes

File: src/audio/component.c
Comment (lyakh) [inline src/audio/component.c:520]: does this all new code really belong to component.c?

File: src/audio/base_fw.c
Comment (lyakh) [inline src/audio/base_fw.c:467]: return codes

File: src/audio/base_fw.c
Comment (lyakh) [inline src/audio/base_fw.c:451]: return codes

File: src/debug/telemetry/telemetry.c
Comment (lyakh) [inline src/debug/telemetry/telemetry.c:245]: 0 is the module ID of the base firmware, right? That's what you mean? Then what does the core ID have to do with module IDs?

File: src/audio/component.c
Comment (lyakh) [inline src/audio/component.c]: this code comes from an earlier commit in this PR, right? Why not use the final version immediately?

File: src/debug/telemetry/telemetry.c
Comment (lyakh) [inline src/debug/telemetry/telemetry.c:366]: codes

File: src/debug/telemetry/telemetry.c
Comment (tobonex) [reply src/debug/telemetry/telemetry.c:72]: Ah, I see. This is used in comp_new and it seems that those can be executed on other cores so I should just add a spinlock for good measure. To be fair, as this code is based on Zephyr's bitarray, it would be a better idea to just add occupied bit count to bitarray in Zephyr, as it already has spinlocks and everything. Getting this accepted may take a while though. Might need to update this later.

File: src/include/sof/debug/telemetry/telemetry.h
Comment (tobonex) [reply src/include/sof/debug/telemetry/telemetry.h:18]: Moved to kconfig as SOF_TELEMETRY_PERFORMANCE_MEASUREMENTS

File: src/audio/component.c
Comment (tobonex) [reply src/audio/component.c]: moved to Kconfig as SOF_TELEMETRY_PERFORMANCE_MEASUREMENTS

File: src/audio/component.c
Comment (tobonex) [reply src/audio/component.c:520]: Well... it doesn't need to be here, but those were the methods in "component" class in reference FW.

File: src/debug/telemetry/telemetry.c
Comment (tobonex) [reply src/debug/telemetry/telemetry.c:264]: In reference FW it was basefw_id + core_id, which effectively is just core id. I also have no idea why it was written like this. This record contains total cycles used by specific core. I'll just remove this comment.

File: src/audio/component.c
Comment (lyakh) [inline src/audio/component.c:573]: parentheses

File: src/audio/component.c
Comment (lyakh) [inline src/audio/component.c:567]: in which cases is `ROUND_UP` not defined?

File: src/audio/component.c
Comment (lyakh) [inline src/audio/component.c:582]: why is this called `source_src_cfg`? What source and what src is it referring to?

File: app/boards/intel_adsp_ace20_lnl.conf
Comment (lyakh) [inline app/boards/intel_adsp_ace20_lnl.conf:43]: with these changes performance measurements would be unconditionally enabled on MTL and LNL? Is this desirable?

File: src/audio/component.c
Comment (lyakh) [inline src/audio/component.c:536]: `return measured_cpc > dev->cpc;`

File: src/audio/component.c
Comment (lyakh) [inline src/audio/component.c:546]: how often does this run? If this runs every millisecond, then in a couple of days this will overflow and the division below will divide by zero

File: src/audio/component.c
Comment (tobonex) [reply src/audio/component.c:567]: There are some builds that don't recognize this. It is defined in Zephyr so I 'm guessing those builds don't use Zephyr (not sure which one it was). This makes it awkward, as I'd rather just use the macro , but it seems I need to expand it anyway because builds will fail otherwise.

File: src/audio/component.c
Comment (tobonex) [reply src/audio/component.c:546]: Yeah it may as well be every ms. Doubt it will be used for a couple of days , but still I should add some check here, true.

File: src/include/sof/debug/telemetry/telemetry.h
Comment (kv2019i) [inline src/include/sof/debug/telemetry/telemetry.h:23]: Hmm, but this is now very Intel specific, should this be called something else than debug/telemtry? debug/intel-telemetry.h?

File: src/debug/telemetry/telemetry.c
Comment (kv2019i) [inline src/debug/telemetry/telemetry.c:183]: I think the file is pretty general otherwise, but this use of the memory window is Intel specific. This itself is ok, but I think it's a problem to call this "SOF debug telemetry" which suggests a generic mechanism, while in practise this only works on platforms that provide a very specific memory window structure (described in Zephyr soc/xtensa/intel_adsp/common/include/adsp-debug-window.h ).

File: src/debug/telemetry/Kconfig
Comment (kv2019i) [inline src/debug/telemetry/Kconfig:20]: So not sure what to do. This text again suggests this is a generic capability, while in practise this can be only used on platforms that implement the adsp-debug.window.h interface (i.e. only Intel platforms now).

File: src/audio/component.c
Comment (lyakh) [inline src/audio/component.c:543]: in `comp_update_ibs_obs_cpc()` in an error case you set `ibs` to 1, which presumably doesn't produce correct performance results. Maybe it's netter to set it to 0 there to skip all the calculations here?

File: src/audio/component.c
Comment (lyakh) [inline src/audio/component.c:553]: but like this after such an overflow they'll stay incorrect? Can we reset `total_cycles_consumed` to start producing valid data again?

File: src/debug/telemetry/telemetry.c
Comment (lyakh) [inline src/debug/telemetry/telemetry.c:373]: nitpicking: don't think `size_t` is a good type for core IDs or for generic cycle indexing below. `size_t` is mostly for byte counts

File: src/audio/component.c
Comment (tobonex) [reply src/audio/component.c:543]: Good idea.

File: src/audio/component.c
Comment (lyakh) [inline src/audio/component.c:577]: don't think you need to divide by 1000 now any more, `SOF_DIV_ROUND_UP()` already does it. And after you've removed it please also drop parentheses

File: src/debug/telemetry/telemetry.c
Comment (lyakh) [inline src/debug/telemetry/telemetry.c:321]: another `size_t`

**Review by lyakh — CHANGES_REQUESTED** [review]

**Comment by tobonex** [comment]
UPDATE:
In ref. FW, this functionality was in separate file so I moved all of this to new files. Now this functionality can be disabled regardless of telemetry as .c files are included conditionally in CMakeLists. Also telemetry.c was getting a bit bloated.

File: src/include/sof/debug/telemetry/telemetry.h
Comment (tobonex) [reply src/include/sof/debug/telemetry/telemetry.h]: Moved new functionality to separate files. I'm guessing we still want those to be called intel-performance-monitor, and intel-telemetry? Do we want to change the CONFIG_SOF_TELEMETRY and CONFIG_SOF_TELEMETRY_PERFORMANCE_MEASUREMENTS names too?

File: src/audio/component.c
Comment (tobonex) [reply src/audio/component.c:578]: The divide is part of the equation, it's correct

File: src/audio/component.c
Comment (tobonex) [reply src/audio/component.c:554]: Hm, we can either do that, or zero everything and set a flag to stop the measurement to further indicate that something went wrong.

File: src/debug/telemetry/Kconfig
Comment (tobonex) [reply src/debug/telemetry/Kconfig:20]: Changed the text, should we change the basic telemetry one as well?

File: src/audio/component.c
Comment (lyakh) [reply src/audio/component.c:578]: Here's the difference to your previous version:
Screenshot_20240704_083619
i.e. basically you replaced `(fmt.sampling_frequency + 999) / 1000` with `SOF_DIV_ROUND_UP(fmt.sampling_frequency, 1000) / 1000` and that is wrong, because `SOF_DIV_ROUND_UP()` is defined as
```
#define SOF_DIV_ROUND_UP(val, div) (((val) + (div) - 1) / (div))
```
and therefore already performs the division by 1000 for you. So, either your original formula was wrong and you actually wanted to divide by 1000000 and not 1000, or this your update is wrong.

Comment (tobonex) [reply src/audio/component.c:578]: The original formula was wrong, yes. I should've mentioned it, sorry.

Comment (kv2019i) [reply src/audio/component.c:578]: I think @tobonex the formula is not correct now:
- 16bit stereo 48000Hz example -> sample group size in bytes 4
- this function would return now: 4 * (48000 / 1000) / 1000 = 0.912 = 0

So I think this should be:
```
return get_sample_group_size_in_bytes(fmt) *
			SOF_DIV_ROUND_UP(fmt.sampling_frequency, 1000);
```

Which gives you 4 * 48000 / 1000 = 192 bytes (which is correct. I think you missed SOF_DIV_ROUND_UP does actually do the dvision.

**Review by kv2019i — COMMENTED** [review]
This starts to be ready for merge. I think we could have a better tree-wide mechanism to identify features built on top of the ADSP_MW/DW (memory window or debug window) interface. But upon second thought, I don't want start renaming the files yet, let's wait whether we can make this generic for other platforms, and if not, make the names more specific later. 

There is one issue w.r.t. 1ms-in-bytes calculation, can you @tobonex check that? Otherwise I'm good with this.

File: src/audio/component.c
Comment (tobonex) [reply src/audio/component.c:578]: Ah, ok i get it now, ROUND_UP_DIV adds division COMPARED to ROUND_UP. I'll fix.

**Review by kv2019i — APPROVED** [review]

**Comment by kv2019i** [comment]
@marcinszkudlinski @abonislawski @tmleman @softwarecki @lgirdwood @lyakh ... this one is waiting for another review, otherwise ready to go.

**Review by lgirdwood — APPROVED** [review]

**Pull request merged**
